Add Polars DataFrame Support for Dataset

ragyabraham opened the pull request:
This PR enhances the library's flexibility by allowing users to leverage Polars DataFrames for training purposes.

### Checklist
- [x] Executed `run-checks all` script
- [x] Updated documentation to reflect changes in this PR

### Related Issues/PRs
Fixes #2015 

### Changes
- Added ability to create a burn dataset directly from a Polars DataFrame for training
- Added a deserializer to deserialize df row into a user provided struct.
- Implemented `dataframe.rs` module in the `burn-dataset` crate under the `dataframe` feature flag
- Updated dataset document

### Testing
- Added additional unit tests to cover new functionality

antimora commented on crates/burn-dataset/Cargo.toml:
```diff
@@ -45,6 +45,7 @@ gix-tempfile = { workspace = true, optional = true }
 globwalk = { workspace = true, optional = true }
 hound = { workspace = true, optional = true }
 image = { workspace = true, optional = true }
+polars = {version =  "0.41.3", features = ["polars-io"] }
```
I would make this dependency as optional, so people using dataset won't have to download polars.

antimora commented on crates/burn-dataset/src/dataset/mod.rs:
```diff
@@ -6,6 +6,7 @@ mod iterator;
 #[cfg(any(feature = "sqlite", feature = "sqlite-bundled"))]
 mod sqlite;
 mod window;
+mod dataframe;
```
We will need to put behind a feature flag.

ragyabraham replied:
yah good point

antimora reviewed (changes requested):
I have a few inlined comments and a question.

Polars has `serde` (feature) support for dataframes (see [code](https://github.com/pola-rs/polars/blob/main/crates/polars/Cargo.toml#L67-L75)). Probably you can bypass using JSON to deserialize row altogether. Have you looked into this?

ragyabraham commented:
> I have a few inlined comments and a question.
> 
> Polars has `serde` (feature) support for dataframes (see [code](https://github.com/pola-rs/polars/blob/main/crates/polars/Cargo.toml#L67-L75)). Probably you can bypass using JSON to deserialize row altogether. Have you looked into this?

Hey @antimora, I did, but I found that the serialize method is not fully implemented in polars yet. I've opened an [issue](https://github.com/pola-rs/polars/issues/17678) on the Polars repo.

antimora commented:
> > I have a few inlined comments and a question.
> > Polars has `serde` (feature) support for dataframes (see [code](https://github.com/pola-rs/polars/blob/main/crates/polars/Cargo.toml#L67-L75)). Probably you can bypass using JSON to deserialize row altogether. Have you looked into this?
> 
> Hey @antimora, I did, but I found that the serialize method is not fully implemented in polars yet. I've opened an [issue](https://github.com/pola-rs/polars/issues/17678) on the Polars repo.

If I understood correctly not all column types supported? Can we support a subset of possible types? This still would be highly useful. Otherwise it seems one needs to encode row into JSON, which won't much more useful than having json row encoded lines in a plain text file. Or maybe I am misunderstanding the use of JSON here.

ragyabraham commented:
> > > I have a few inlined comments and a question.
> > > Polars has `serde` (feature) support for dataframes (see [code](https://github.com/pola-rs/polars/blob/main/crates/polars/Cargo.toml#L67-L75)). Probably you can bypass using JSON to deserialize row altogether. Have you looked into this?
> > 
> > 
> > Hey @antimora, I did, but I found that the serialize method is not fully implemented in polars yet. I've opened an [issue](https://github.com/pola-rs/polars/issues/17678) on the Polars repo.
> 
> If I understood correctly not all column types supported? Can we support a subset of possible types? This still would be highly useful. Otherwise it seems one needs to encode row into JSON, which won't much more useful than having json row encoded lines in a plain text file. Or maybe I am misunderstanding the use of JSON here.

it seems that the [pull request](https://github.com/pola-rs/polars/pull/17679) to fix the todo! issue in polars might get through. if that's the case then we can just utilise that. will be cleaner

antimora commented:
> > > > I have a few inlined comments and a question.
> > > > Polars has `serde` (feature) support for dataframes (see [code](https://github.com/pola-rs/polars/blob/main/crates/polars/Cargo.toml#L67-L75)). Probably you can bypass using JSON to deserialize row altogether. Have you looked into this?
> > > 
> > > 
> > > Hey @antimora, I did, but I found that the serialize method is not fully implemented in polars yet. I've opened an [issue](https://github.com/pola-rs/polars/issues/17678) on the Polars repo.
> > 
> > 
> > If I understood correctly not all column types supported? Can we support a subset of possible types? This still would be highly useful. Otherwise it seems one needs to encode row into JSON, which won't much more useful than having json row encoded lines in a plain text file. Or maybe I am misunderstanding the use of JSON here.
> 
> it seems that the [pull request](https://github.com/pola-rs/polars/pull/17679) to fix the todo! issue in polars might get through. if that's the case then we can just utilise that. will be cleaner

Yeah I think we should definitely use Dataframe's native serialization/deserialization methods. Otherwise this Dataset will be little of use. 

A cool use case I envision as follows:

1. One can open any Polars datasource (CVS, JSON, binaries, SQL Dbs, etc) into Dataframe. 
2.  Apply [dataframe filters ](https://docs.pola.rs/api/python/stable/reference/dataframe/api/polars.DataFrame.filter.html) to filter (
3. Define struct for a row
4. Wrap with Dataframe Dataset (can also be passed to SqliteDataset or MemoryDataset to consume and cache locall)
5. Pass it to training/testing

BTW, your PR is waiting to fix this. Not sure if you noticed:

<img width="1059" alt="image" src="https://github.com/user-attachments/assets/de92c3e3-47cb-4a09-a9e4-fd73c526e3d0">

antimora commented on crates/burn-dataset/src/dataset/dataframe.rs:
```diff
@@ -0,0 +1,228 @@
+    }
+
+    #[rstest]
+    fn len(train_dataframe: DataframeSample) {
```
`train_dataframe` is confusing name. Use ds for dataset.

antimora commented on crates/burn-dataset/src/dataset/dataframe.rs:
```diff
@@ -0,0 +1,228 @@
+}
+
+impl<I> DataframeDataset<I> {
+     /// Initialise a DataframeDataset from a polars dataframe
```
There is a leading space before '///'

antimora commented on crates/burn-dataset/src/dataset/dataframe.rs:
```diff
@@ -0,0 +1,228 @@
+
+impl<I> DataframeDataset<I> {
+     /// Initialise a DataframeDataset from a polars dataframe
+    pub fn from_dataframe(df: DataFrame) -> Result<Self> {
+        let len = df.height();
+        Ok(DataframeDataset {
+            dataframe: df,
+            len,
+            phantom: PhantomData,
+        })
+    }
```
Can we also add `from_lazy_dataframe`?

antimora commented on crates/burn-dataset/src/dataset/dataframe.rs:
```diff
@@ -0,0 +1,228 @@
+/// each instance of this struct will represent a single polars dataframe.
+/// this struct provides basic Dataset functionality to a polars dataframe, so that they are able
+/// to natively be used for training/validation.
+/// Currently there is no native way to deserialize a polars row, so a trial and error approach
+/// has been taken. This could very well change in the future.
```
Once you have the native deserializer working, this should be updated.

antimora commented on crates/burn-dataset/src/dataset/dataframe.rs:
```diff
@@ -0,0 +1,228 @@
+
+pub type Result<T> = core::result::Result<T, DataframeDatasetError>;
+
+/// Custom error type for DataframeDataset type
+#[derive(thiserror::Error, Debug)]
+pub enum DataframeDatasetError {
+    /// IO related error.
+    #[error("IO error: {0}")]
+    Io(#[from] io::Error),
+    /// Serde related error.
+    #[error("Serde error: {0}")]
+    Serde(#[from] rmp_serde::encode::Error),
+    /// Polars related error
+    #[error("Polars error: {0}")]
+    PolarsError(#[from] PolarsError),
+    /// Any other error.
+    #[error("{0}")]
+    Other(&'static str),
+}
+
+impl From<&'static str> for DataframeDatasetError {
+    fn from(s: &'static str) -> Self {
+        DataframeDatasetError::Other(s)
+    }
+}
```
I am not sure if these errors are utilized anywhere. Can you make sure you use or remove them?

ragyabraham commented:
> We spoke offline about using Polar's serde feature to deserialize a row. I am not sure if new code wasn't pushed (that you mentioned working), or there is still misunderstanding.

The code that I showed you working is the code in this PR as it stands

ragyabraham commented on crates/burn-dataset/src/dataset/dataframe.rs:
```diff
@@ -0,0 +1,228 @@
+                    false => { None }
+                }
+            },
+            Err(e) => {
+                println!("An error occurred while checking if a Polars row is serializable: Error: {:?}", e);
+                None
+            }
+        }
+            },
+            Err(e) =>{
+                println!("An error occurred while getting row from polars dataframe. Error: {}", e);
+                None
+            }
+        }
```
Sure

antimora commented:
OK. I removed serde_json and implemented the native deserializer. Please finish up and clean up the code. Also please add more tests and data types. 

One test is passing: 

```
[burn-dataset]% cargo test dataframe --features dataframe
    Finished `test` profile [unoptimized] target(s) in 0.12s
     Running unittests src/lib.rs (/Users/dilshod/Projects/burn/target/debug/deps/burn_dataset-c350f307979d05bd)

running 1 test
test dataset::dataframe::tests::test_dataframe_dataset ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 25 filtered out; finished in 0.00s

```

ragyabraham commented:
> OK. I removed serde_json and implemented the native deserializer. Please finish up and clean up the code. Also please add more tests and data types.
> 
> One test is passing:
> 
> ```
> [burn-dataset]% cargo test dataframe --features dataframe
>     Finished `test` profile [unoptimized] target(s) in 0.12s
>      Running unittests src/lib.rs (/Users/dilshod/Projects/burn/target/debug/deps/burn_dataset-c350f307979d05bd)
> 
> running 1 test
> test dataset::dataframe::tests::test_dataframe_dataset ... ok
> 
> test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 25 filtered out; finished in 0.00s
> ```

@antimora I've added support for `LazyFrame`, added support to deserialize a few more datatypes and added some tests. So far, there's no serialization for complex types, but this is a good starting point. Let me know what you think and if there's anything urgent I need to fix/add

Two tests are passing

![image](https://github.com/user-attachments/assets/e85cef3f-af5f-4144-b4e2-70297c74a80d)

antimora commented on crates/burn-dataset/src/dataset/dataframe.rs:
```diff
@@ -0,0 +1,390 @@
+    }
+}
+
+/// Enum to hold either a DataFrame or a LazyFrame
+pub enum Frame {
+    /// Represents a Polars Dataframe
+    Data(DataFrame),
+    /// Represents a Polars LazyFrame
+    Lazy(LazyFrame),
+}
```
Let's remove Lazy for now to get the approval quicker (I suggested Lazy as well but it appears fetching records more complex and currently you're just returning None).

antimora commented on crates/burn-dataset/src/dataset/dataframe.rs:
```diff
@@ -0,0 +1,390 @@
+
+/// Dataset implementation for Polars Frame (DataFrame or LazyFrame)
+pub struct DataframeDataset<I> {
+    frame: Frame,
```
Replace with Dataframe

antimora commented on crates/burn-dataset/src/dataset/dataframe.rs:
```diff
@@ -0,0 +1,390 @@
+                    phantom: PhantomData,
+                })
+            },
+            Frame::Lazy(lazy) => {
+                let schema = lazy
+                    .clone()
+                    .schema()
+                    .map_err(
+                        |e| DataframeDatasetError::Other(e.to_string()))?;
+                // LazyFrame doesn't know the length upfront so set to zero initially 
+                let len = 0;
+                let field_names = extract_field_names::<I>();
+
+                let column_name_mapping = field_names
+                    .iter()
+                    .filter_map(|name| {
+                        schema
+                            .iter()
+                            .position(|(col_name, _)| col_name == name)
+                    })
+                    .collect::<Vec<_>>();
+                Ok(DataframeDataset {
+                    frame: Frame::Lazy(lazy),
+                    len,
+                    column_name_mapping,
+                    phantom: PhantomData,
+                })
+            },
```
Lets remove Lazy here.

antimora commented on crates/burn-dataset/src/dataset/dataframe.rs:
```diff
@@ -0,0 +1,390 @@
+        }
+    }
+
+    /// Create a new DataframeDataset from a Polars LazyFrame
+    pub fn from_lazyframe(lf: LazyFrame) -> Result<Self> {
+        Self::new(Frame::Lazy(lf))
+    }
```
Remove lazy.

antimora commented on crates/burn-dataset/src/dataset/dataframe.rs:
```diff
@@ -0,0 +1,390 @@
+        Self::new(Frame::Data(df))
+    }
+
+    /// Collect the LazyFrame into a DataFrame and get the length
+    pub fn collect_and_get_len(&mut self) -> Result<usize> {
+        if let Frame::Lazy(lazy) = &self.frame {
+            let df = lazy.clone().collect().map_err(
+                |e|
+                DataframeDatasetError::Other(e.to_string())
+            )?;
+            self.len = df.height();
+            self.frame = Frame::Data(df);
+        }
+        Ok(self.len)
+    }
```
Remove lazy

antimora commented on crates/burn-dataset/src/dataset/dataframe.rs:
```diff
@@ -0,0 +1,390 @@
+            let mut deserializer = RowDeserializer::new(&row, &self.column_name_mapping);
+            I::deserialize(&mut deserializer).ok()
+        } else {
+            None // Or handle LazyFrame differently if needed
```
This part appears complex and currently not complete as it stands.

antimora commented on crates/burn-dataset/src/dataset/dataframe.rs:
```diff
@@ -0,0 +1,390 @@
+            AnyValue::Float64(f) => visitor.visit_f64(*f),
+            AnyValue::Date(i) => visitor.visit_i32(*i),
+            AnyValue::String(s) => visitor.visit_string(s.to_string()),
+            AnyValue::Binary(b) => {
+                let v: Vec<usize> = b.iter().map(|&x| x as usize).collect();
+                visitor.visit_seq(de::value::SeqDeserializer::new(v.into_iter()))
```
I think for Binary you want to use u8 (bytes) instead of usize.

antimora commented on crates/burn-dataset/src/dataset/dataframe.rs:
```diff
@@ -0,0 +1,390 @@
+        DataFrame::new(vec![s1, s2, s0, s3, s4]).unwrap()
+    }
+
+    fn create_test_lazyframe() -> LazyFrame {
+        let s0 = Series::new("a", &[1, 2, 3, ]);
+        let s1 = Series::new("b", &[true, false, true]);
+        let s2 = Series::new("c", &[1.1, 2.2, 3.3]);
+        let s3 = Series::new("d", &["Boo", "Boo2", "Boo3"]);
+        let s4 = Series::new("e", &[
+            vec![1, 2, 3],
+            vec![4, 5, 6],
+            vec![7, 8, 9]
+        ]);
+        DataFrame::new(vec![s1, s2, s0, s3, s4]).unwrap().lazy()
+    }
```
Remove lazy

antimora commented on crates/burn-dataset/src/dataset/dataframe.rs:
```diff
@@ -0,0 +1,390 @@
+        );
+    }
+
+    #[test]
+    fn test_lazyframe_dataset() {
+        let df = create_test_lazyframe();
+        let mut dataset: DataframeDataset<TestData> = DataframeDataset::from_lazyframe(df).unwrap();
+        dataset.collect_and_get_len().unwrap();
+
+        assert_eq!(dataset.len(), 3);
+        assert_eq!(dataset.is_empty(), false);
+
+        let item = dataset.get(1).unwrap();
+        assert_eq!(
+            item,
+            TestData {
+                a: 2,
+                b: false,
+                c: 2.2,
+                d: "Boo2".to_string(),
+                e: vec![4, 5, 6],
+            }
+        );
+
+        let item = dataset.get(2).unwrap();
+
+        assert_eq!(
+            item,
+            TestData {
+                a: 3,
+                b: true,
+                c: 3.3,
+                d: "Boo3".to_string(),
```
Remove lazy

antimora commented on crates/burn-dataset/src/dataset/dataframe.rs:
```diff
@@ -0,0 +1,390 @@
+        );
+    }
+
+    // TODO test non existing struct fields
```
Missing this TODO. You should test have some data that's not present in your struct fields.

antimora reviewed (changes requested):
Let's remove Lazy for now. I thought it would be less complex and quick to add it but as it stands it's not complete and I don't want to burden you with this feature, sorry.

We need a couple minor/quick fixes. It should be ready to merge afterwards.

Also please rebase your branch to get the latest lock file.

antimora commented:
@laggui @nathanielsimard I took over this PR to complete it. I made final changes and ready for your review.

nathanielsimard reviewed (approved):
LGTM from my end. The deserializer logic is a bit complex, but the tests comparing with a vector dataset are reassuring. And I guess it's necessary with Polar, though I'm not familiar with the library.

antimora reviewed (approved):
Since I made the actual changes and fixes, I approve it.

wangjiawen2013 commented:
Good!
